File: apps/party-time-backend/src/main/java/com/partytime/jpa/mapper/AddressMapper.java
package com.partytime.jpa.mapper;

import com.partytime.api.dto.address.AddressDTO;
import com.partytime.jpa.entity.Address;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.util.Optional;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class AddressMapper {

    public static AddressDTO map(Address address) {
        return AddressDTO.builder()
            .addressLine(address.getAddressLine())
            .addressLineAddition(address.getAddressLineAddition())
            .zip(address.getZip())
            .city(address.getCity())
            .country(address.getCountry())
            .build();
    }

    public static Address map(AddressDTO address) {
        return Address.builder()
            .addressLine(address.getAddressLine())
            .addressLineAddition(address.getAddressLineAddition())
            .zip(address.getZip())
            .city(address.getCity())
            .country(address.getCountry())
            .build();
    }

    public static Address create(String addressLine,
                                 String addressLineAddition,
                                 String zip,
                                 String city,
                                 String country) {
        return Address.builder()
            .addressLine(addressLine)
            .addressLineAddition(addressLineAddition)
            .zip(zip)
            .city(city)
            .country(country)
            .build();
    }

    public static String prettyPrint(Address address) {
        return address.getAddressLine()
            + Optional.ofNullable(address.getAddressLineAddition())
            .map(s -> "\n" + s)
            .orElse("")
            + "\n" + address.getZip() + " " + address.getCity()
            + "\n" + address.getCountry();
    }

}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/jpa/mapper/AddressMapper.kt
package com.partytime.jpa.mapper

import com.partytime.api.dto.address.AddressDTO
import com.partytime.jpa.entity.Address

/**
 * Transforms an [Address] entity into a (serializable) [AddressDTO]
 *
 * @receiver transformation subject
 * @return transformation result, ready for serialization and transmission to the client
 */
fun Address.toAddressDTO(): AddressDTO = AddressDTO(
    addressLine,
    addressLineAddition,
    zip,
    city,
    country
)

/**
 * Formats an [Address] entity in a human-readable format.
 *
 * @receiver formatting subject
 * @return address block
 */
fun Address.toMultiLineString(): String = listOfNotNull(
    addressLine,
    addressLineAddition?.let { "\n$it" },
    "$zip $city",
    country
).joinToString("\n")
